### DISCUSSION TITLE
Support not full graph for VRP (performance reason)

### DISCUSSION THREAD

**Discussion by bguban** [google/or-tools#3397] [Routing (and legacy CP) questions]
I build routes for merchandisers who walks between shopping points and check the goods on the shelves. The points are pretty dense (it can be just a few minutes to walk between) but there are a lot of them (the entire city). The merchandiser should spend 20 minutes in a single point and has 6 hours working day. So logically there is no sense even to try to add points that are on opposite sides of the city into the single route. But or-tools tries to use all the points that slows down the process a lot.

As a solution, we can use a distance hash instead of the distance matrix. Something like: {node1: {node2: distance, node20: distance}, node2: {node1: distance, node3: distance}}. Or talking in words of or-tool we can have a callback that would receive a node and return back an array of nodes that can directly communicate with the given node. Also, I guess it would be useful for some algorithms to know what nodes are closest to the given node (sorted array) to take the closest one instead of checking all possible nodes.

**Comment by Regista6** [comment]
Maybe you can use some heuristic to calculate the initial solution quickly and then use something like guided local search to improve the solution.
Or you can cluster the points and then and then solve the problem for each individual cluster.

**Comment by bguban** [comment]
Currently clustering is the best solution I found but it has problems on cluster borders. The pink arrow (in the bottom) shows such case. The route from the left cluster goes up instead of collecting closest points from the central cluster.
<img width="234" alt="Screen Shot 2022-07-11 at 12 14 43 PM" src="https://user-images.githubusercontent.com/1564376/180612347-6221bf53-92b1-46f9-a29a-a9b8183fc2fd.png">

I can build an initial route but any way the guided local search must check all the nodes to find the closest that is O(n**2) complexity. And as I know (I might be wrong) the guided local search doesn't really consider only closest nodes. It takes all possible nodes.

**Reply by Regista6** [reply]
You can set the time limit of guided local search(gls). Worth a try in my opinion (if you haven't already). As far as I know, gls is just a metaheuristic that guides the underlying local search when it is stuck in a local minimum. The efficiency of a particular local search method (inter-route, intra-route) depends on the size of its neighborhood. Care is usually taken so that the neighborhoods are not so large that it becomes inefficient to generate the solutions and evaluate them!.

**Reply by bguban** [reply]
I have a time limit (otherwise the process would be infinite) but it just stops the process when the limit is reached.
> The efficiency of a particular local search method (inter-route, intra-route) depends on the size of its neighborhood.

that is exactly what I'm trying to reach. Lets' imagine we are going to apply the 2Opt operation to build a new neighborhood. By default, all the paths in the route are candidates because we have a full distance matrix (any node is reachable from any other node) that is a huge amount of combinations. For intra-route, the number of operations is even bigger because the candidates are all nodes from all other routes. The solution is to filter out far candidates.

In my task, I have nearly 6000 nodes (6K**2 = 32M matrix size). Nodes are densely distributed. There is no reason even to try to bind nodes on different sides of the city. Ideally, I would like to allow a node to bind to 20 closest nodes (6K * 20 = 120K) that are 300 times smaller.
